pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/TripleLongPriorityQueue.java

```diff
@@ -152,6 +174,21 @@ private void increaseCapacity() {
         buffer.capacity(this.capacity * TUPLE_SIZE);
     }
 
+    private void shrinkCapacity() {
+        if (capacity > initialCapacity &&  size < capacity * shrinkFactor) {
+            int decreasingSize = (int) (capacity * shrinkFactor * RESERVATION_FACTOR);
+            if (decreasingSize <= 0) {
+                return;
+            }
+            if (capacity - decreasingSize <= initialCapacity) {
+                this.capacity = initialCapacity;
+            } else {
+                this.capacity = capacity - decreasingSize;
+            }
+            buffer.capacity(this.capacity * TUPLE_SIZE);
```

Comment:
I think this will truncate the buffer, but it will not release the underlying memory back to the pool: https://netty.io/4.1/api/io/netty/buffer/ByteBuf.html#capacity-int- 

We'd probably have to allocate a new buffer and copy the content into it.